src/common/buffer.cc

```diff
@@ -1645,7 +1645,6 @@ static ceph::spinlock debug_lock;
         curbuf = _buffers.erase_after(curbuf_prev);
 	_carriage->set_offset(_carriage->offset() + _carriage->length());
 	_carriage->set_length(0);
-	_buffers.push_back(*_carriage);
```

Comment:
Well, if we really need to drop this line, the comment above should be reworked as well. Also, I'm worried the patch introduces a memory leak as `erase_after()` doesn't dispose the node.

In the meantime, I joining the folks asking for more descriptive commit message.

Reply:
> Well, if we really need to drop this line, the comment above should be reworked as well. Also, I'm worried the patch introduces a memory leak as `erase_after()` doesn't dispose the node.
> 
> In the meantime, I joining the folks asking for more descriptive commit message.

I'm not sure about the solution, but the problem does exist. see the issue describe by Jianpeng

Reply:
@CongMinYin, @majianpeng: I'm not questioning that at all. :-) Just arguing about the way how we fix it.

>  a)bl has a ptr which length is zero.

This is still valid and can (rarely) happen because of how `_carriage` and the non-pointer `append()` variants of `contiguous_appender` interact:

```cpp
    class contiguous_appender {
      // ...
      void append(const bufferlist& l) {
        if (deep) {
          // ...
        } else {
          flush_and_continue();
          bl.append(l);
          space = bl.obtain_contiguous_space(0);
          out_of_band_offset += l.length();
        }
      }
```

My impression is we need to take a look on `rebuild_aligned_size_and_memory()`. I can help with that.

Reply:
Also, take a look on the `BufferListIterator,iterate_with_empties` testcase.

Reply:
We also look forward to discussing with you and seeking the most appropriate solution.

Reply:
OK, I think the fix could look like this one:

```diff
diff --git a/src/common/buffer.cc b/src/common/buffer.cc
index 548abc458fd..9c76cde6eed 100644
--- a/src/common/buffer.cc
+++ b/src/common/buffer.cc
@@ -1258,8 +1258,12 @@ static ceph::spinlock debug_lock;
             buffer::create_aligned(unaligned._len, align_memory)));
         had_to_rebuild = true;
       }
-      _buffers.insert_after(p_prev, *ptr_node::create(unaligned._buffers.front()).release());
-      _num += 1;
+      if (unaligned.get_num_buffers()) {
+        _buffers.insert_after(p_prev, *ptr_node::create(unaligned._buffers.front()).release());
+        _num += 1;
+      } else {
+        // a bufferlist containing only 0-length bptrs is rebuild as empty
+      }
       ++p_prev;
     }
     return had_to_rebuild;
```

Reply:
Thank you for your solution. I have updated the code and commit information of PR.